### ISSUE TITLE
How to add support for new boards

### ISSUE THREAD

**Issue by ggsnpdsn** [ZRouter/ZRouter#44]
The current boards that support atheros are outdated. I want to add some boards that support the qca series, but I don't know how to do it. Is there any reference or instructions?

**Comment by rayddteam** [comment]
You can find some instructions here:
https://zrouter.org/projects/zrouter/wiki/For_developers
But it can be a bit outdated too :)
Main problem here is in that FreeBSD team in progress of removing MIPS support, due to lack of maintainers.
I've just created for of FreeBSD tree, so when time permit, we will restore MIPS support bits in that tree.
https://github.com/ZRouter/freebsd-src
Welcome to community!
don't expect quick answers, but we will try to help as soon as able.
Thanks!

**Comment by ggsnpdsn** [comment]
> You can find some instructions here: https://zrouter.org/projects/zrouter/wiki/For_developers But it can be a bit outdated too :) Main problem here is in that FreeBSD team in progress of removing MIPS support, due to lack of maintainers. I've just created for of FreeBSD tree, so when time permit, we will restore MIPS support bits in that tree. https://github.com/ZRouter/freebsd-src Welcome to community! don't expect quick answers, but we will try to help as soon as able. Thanks!

Thank you very much for your reply, I read the developer guide you posted, just add board.hint and board.mk to the boards directory?
I plan to add a device that uses the qca9563 chip if it works

**Comment by yamori813** [comment]
I think you use clang for cross comple at 13R or later. mips support not work on clang. You need use FreeBSD 12-Stable source tree.

**Comment by ggsnpdsn** [comment]
> 
yes，i am used FreeBSD 13-Stable source on my FreeBSD 13 host，try FreeBSD 12-Stable source later

**Comment by ggsnpdsn** [comment]
Does it not support wireless? I set WITHOUT_WIRELESS=no, but the compilation error is as follows:

-c /usr/home/jhon/ZRouter/ZRouter/FreeBSD12/tools/tools/ath/athani/main.c -o main.o
*** Error code 1

Stop.
make[5]: stopped in /usr/home/jhon/ZRouter/ZRouter/FreeBSD12/tools/tools/ath/athani
*** Error code 1

Stop.
make[4]: stopped in /usr/home/jhon/ZRouter/ZRouter/FreeBSD12/tools/tools/ath
*** Error code 1

Stop.
make[3]: stopped in /usr/home/jhon/ZRouter/ZRouter/FreeBSD12
*** Error code 1

Stop.
make[2]: stopped in /usr/home/kyslink/ZRouter/ZRouter/FreeBSD12

**Comment by yamori813** [comment]
Yes. I never use WiFi on FreeBSD. In Japane not have examination is illegal.

**Comment by ggsnpdsn** [comment]
> Yes. I never use WiFi on FreeBSD. In Japane not have examination is illegal.

If I want to use wifi, which entry do I need to modify the code or debug, not too familiar with FreeBSD code

**Comment by yamori813** [comment]
I think only few on Wifi because of atheros gay mentenance wifi code on FreeBSD. But Wifi is very complex. I think not good that heaby use wifi on FreeBSD.

**Comment by ggsnpdsn** [comment]
What is the main difference between SMALL_ and XSMALL_?

**Comment by yamori813** [comment]
xSMALL_ is minimum set. SMALL_ add more commands and files.

**Comment by ggsnpdsn** [comment]
Take WN-G300R as an example:
Where are the variables defined in boards/IOData/WN-G300R/board.hints referenced? What are the rules and requirements for variable definition?
Are the compile options in boards/IOData/WN-G300R/board.mk? where the reference takes effect.
Are socs/Atheros/AR9132/soc.mk and socs/Atheros/AR9132/soc.hints written according to the chip manual? Where can variable or parameter name definition rules or references be found?
By the way: After the compilation is completed, is X.zImage the image used to upgrade the router?

**Comment by yamori813** [comment]
I do modify from freebsd srouce tree sys/mips/conf file. I do separate soc and board setting hints and mk file.

zimage file is u-boot image file. You can't upgrade from original firmware. You need u-boot operation flash to zimage file .

**Comment by ggsnpdsn** [comment]
I selected ath and hostapd in the profile, and the compilation produced an error: ZRouter/FREEBSD12/mips.mips/tmp/usr/bin/ld: cannot find -lwpadrivers

How to fix it, I'm not too familiar with this build system

**Comment by yamori813** [comment]
If you want make base station on mips module I suggest use OpenWRT more than FreeBSD.

**Comment by ggsnpdsn** [comment]
I used the profile of IODATA WN G300R as the basis, and modified the flash layout. The image I compiled can already run on wpj342, but the lan interface is blocked, and I cannot access 192.168.0.1. I don’t know where I made a mistake.

**Comment by yamori813** [comment]
Atheros AR9342 is two port ethernet interface. That is arge0 and arge1. One is connect to ethernet swich and other is connect to ethernet swich phy. I seem WPJ342 have external ethernet swich chip. I seem that is athros swich chip. FreeBSD have athros ethernet swich chip code but very confuse. You need configure arge intafece and internal etherswich and external etherswich.
